### PR TITLE
Add support for cmd args

### LINKED ISSUE #22: Support command line options in lib::Initialize()
Node.js supports a lot of CLI flags, many of which could also be useful when running it as a shared library. We should allow setting these (e.g. by passing a string as a parameter) in the Initialize() call.

reference: https://nodejs.org/api/cli.html

### PR DESCRIPTION
**Pull request by luminosuslight**
Working:
- [x] node commandline arguments
- [x] v8 commandline arguments

Closes #22.

### CODE REVIEW

File: src/node.cc [hunk 1]
```diff
@@ -5060,13 +5063,26 @@ void deinitV8() {
...
+  cmd_args->argument_pointer.push_back(const_cast<char*>(cmd_args->argument_data.data() + cmd_args->argument_data.size()));
+  cmd_args->argument_data += program_name;
+  cmd_args->argument_data += char(0x0);
+  for (const auto& s: arguments) {
+    cmd_args->argument_pointer.push_back(const_cast<char*>(cmd_args->argument_data.data() + cmd_args->argument_data.size()));
+    cmd_args->argument_data += s;
+    cmd_args->argument_data += char(0x0);
+  }
+  cmd_args->argc = cmd_args->argument_pointer.size();
+  cmd_args->argv = cmd_args->argument_pointer.data();
```
Comment (cmfcmf) [inline src/node.cc]: Can we move this into a member function (constructor?) of the `CmdArgs` class?

**Comment by justus-hildebrand** [comment]
another Todo: test uv args.

File: src/node.cc (hunk 1 above)
Comment (luminosuslight) [reply src/node.cc]: Good point!

**Comment by luminosuslight** [comment]
It seems like a function `func(std::initializer_list<std::string> l)` needs to be called as `func({"foo", "bar"})` (with curly braces, see http://en.cppreference.com/w/cpp/utility/initializer_list). This means that it makes no difference whether vector or initializer_list is used.

**Comment by cmfcmf** [comment]
> It seems like a function func(std::initializer_list<std::string> l) needs to be called as func({"foo", "bar"}) (with curly braces, see http://en.cppreference.com/w/cpp/utility/initializer_list).

What else did you/we expect?

File: src/node.cc [hunk 2]
```diff
@@ -4993,8 +4993,40 @@ int Start(int argc, char** argv) {
...
+      argument_data += char(0x0);
+    }
+    argc = argument_pointers.size();
+    argv = argument_pointers.data();
+  }
+
   int argc;
   char** argv;
+  std::string argument_data;
+  std::vector<char*> argument_pointers;
```
Comment (cmfcmf) [inline src/node.cc]: These two don't need to be public.

File: src/node.cc [hunk 3]
```diff
@@ -4993,8 +4993,40 @@ int Start(int argc, char** argv) {
 namespace lib {
 
 struct CmdArgs {
+  /**
+   * @brief CmdArgs builds a continuous adjacent char buffer (argv) from a vector of strings
+   * and stores it in the static cmd_args variable
```
Comment (cmfcmf) [inline src/node.cc]: Technically, it's not really a "static" variable, is it? Looks like a simple member variable to me.

Comment (luminosuslight) [reply src/node.cc]: right, I fixed the documentation

File: src/node.cc (hunk 2 above)
Comment (luminosuslight) [reply src/node.cc]: 👍

**Comment by luminosuslight** [comment]
> > It seems like a function func(std::initializer_liststd::string l) needs to be called as func({"foo", "bar"}) (with curly braces, see http://en.cppreference.com/w/cpp/utility/initializer_list).

> What else did you/we expect?

It would be cool if you could call it as func("foo", "bar") instead of with curly braces, but it seems like this is only possible with real variadic functions or recursive templates (which would be hard to maintain in this case).

**Pull request merged**